packages/snaps-utils/src/url.ts

```diff
@@ -0,0 +1,82 @@
+import { type SnapId } from '@metamask/snaps-sdk';
+import { assert } from '@metamask/utils';
+
+import { assertIsValidSnapId, stripSnapPrefix } from './snaps';
+
+export type Authority = 'client' | 'snap';
+
+export const CLIENT_PATHS = ['/'];
+
+export const SNAP_PATHS = ['/home'];
+
+/**
+ * Parse a url string to an object containing the authority, path and Snap id (if snap authority).
+ * This also validates the url before parsing it.
+ *
+ * Note: The Snap id returned from this function should always be validated to be an installed snap.
+ *
+ * @param str - The url string to validate and parse.
+ * @returns A parsed url object.
+ * @throws If the authority or path is invalid.
+ */
+export function parseMetaMaskUrl(str: string): {
+  authority: Authority;
+  snapId?: SnapId;
+  path: string;
+} {
+  const baseErrorMessage = 'Invalid MetaMask url:';
+  const url = new URL(str);
+  const { hostname: authority, pathname: path, protocol } = url;
+  if (protocol !== 'metamask:') {
+    throw new Error(
+      `Unable to parse URL. Expected the protocol to be "metamask:", but received "${protocol}".`,
+    );
+  }
+  if (authority === 'client') {
```

Comment:
Maybe change this into a switch statement for future additions:

```suggestion
  switch (authority) {
    case 'client':
      // ...
      
    case 'snap':
      // ...
      
    default:
      throw new Error('Invalid `metamask://` URL.');
  }
```